<?xml version="1.0"?>
<launch>
  <!-- Hand Control System Launch File -->
  
  <!-- Load configuration -->
  <rosparam file="$(find robotic_arm_controller)/config/robot_config.yaml" command="load"/>
  <rosparam file="$(find robotic_arm_controller)/config/hand_tracking_config.yaml" command="load"/>
  
  <!-- Hand Tracker Node -->
  <node name="hand_tracker_node" pkg="robotic_arm_controller" type="hand_tracker_node.py" output="screen">
    <param name="camera_id" value="0"/>
    <param name="model_complexity" value="1"/>
    <param name="min_detection_confidence" value="0.7"/>
    <param name="min_tracking_confidence" value="0.5"/>
    <param name="smoothing_alpha" value="0.3"/>
    <param name="publish_debug_image" value="true"/>
  </node>
  
  <!-- Coordinate Mapper Node -->
  <node name="coordinate_mapper_node" pkg="robotic_arm_controller" type="coordinate_mapper_node.py" output="screen">
    <param name="camera_x_min" value="0.2"/>
    <param name="camera_x_max" value="0.8"/>
    <param name="camera_y_min" value="0.2"/>
    <param name="camera_y_max" value="0.8"/>
    <param name="camera_z_min" value="-0.1"/>
    <param name="camera_z_max" value="0.1"/>
    
    <param name="robot_x_min" value="0.2"/>
    <param name="robot_x_max" value="0.6"/>
    <param name="robot_y_min" value="-0.3"/>
    <param name="robot_y_max" value="0.3"/>
    <param name="robot_z_min" value="0.1"/>
    <param name="robot_z_max" value="0.5"/>
    
    <param name="flip_x" value="false"/>
    <param name="flip_y" value="true"/>
    <param name="flip_z" value="false"/>
    
    <param name="smoothing_alpha" value="0.2"/>
  </node>
  
  <!-- IK Solver Node -->
  <node name="ik_solver_node" pkg="robotic_arm_controller" type="ik_solver_node.py" output="screen">
    <param name="solver_method" value="numerical"/>
    <param name="use_current_as_seed" value="true"/>
  </node>
  
  <!-- Robot Controller Node -->
  <node name="robot_controller_node" pkg="robotic_arm_controller" type="robot_controller_node.py" output="screen">
    <param name="control_mode" value="position"/>
    <param name="smoothing_alpha" value="0.3"/>
    <param name="max_joint_velocity" value="1.0"/>
  </node>
  
  <!-- Optional: Image view for debug -->
  <node name="debug_image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
    <remap from="image" to="/hand_tracker/debug_image"/>
  </node>

</launch>
